File: tree/verification-validation/BalancedTreeCheck.cpp

class BalancedTreeCheck {
public:
    bool isBalanced(TreeNode* root) {
        if (root == NULL) {
            return true;
        }
        if (abs(getMaxDepth(root->left) - getMaxDepth(root->right)) > 1) {
            return false;
        }
        return isBalanced(root->left) && isBalanced(root->right);
    }

private:
    int getMaxDepth(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }
        return 1 + max(getMaxDepth(root->left), getMaxDepth(root->right));
    }
};

File: tree/verification-validation/BalancedTreeCheck.java

public class BalancedTreeCheck {
    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }
        if (Math.abs(getDepth(root.left) - getDepth(root.right)) > 1) {
            return false;
        }
        return isBalanced(root.left) && isBalanced(root.right);
    }

    private int getDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return 1 + Math.max(getDepth(root.left), getDepth(root.right));
    }
}
